File: server/app/schemas/leaderboard.py
"""Leaderboard bundle schemas."""

from typing import Any

from pydantic import BaseModel


class LeaderboardMetric(BaseModel):
    """Leaderboard metric."""

    hasData: bool
    method: str
    currentValue: int
    keyField: str | None = None
    trendData: list[Any]
    dataPoints: list[Any]
    hover: dict[str, Any]


class LeaderboardMetrics(BaseModel):
    """Leaderboard metrics."""

    totalAttempts: LeaderboardMetric
    highestScoreAvg: LeaderboardMetric
    messagesPerSession: LeaderboardMetric
    personaResponseSeconds: LeaderboardMetric
    timeSpentMinutes: LeaderboardMetric
    improvementRatePerDay: LeaderboardMetric
    perfectScoreCount: LeaderboardMetric
    quickestPassMinutes: LeaderboardMetric


class LeaderboardRow(BaseModel):
    """Leaderboard row."""

    profileId: str
    firstName: str
    lastName: str
    metrics: LeaderboardMetrics


class LeaderboardBundleResponse(BaseModel):
    """Leaderboard bundle response."""

    data: list[LeaderboardRow]

File: client/lib/api/v2/schemas/leaderboard.ts
/**
 * Leaderboard Bundle Schemas
 */

import { z } from "zod";

export const LeaderboardMetricSchema = z.object({
  hasData: z.boolean(),
  method: z.string(),
  currentValue: z.number(),
  keyField: z.string().nullish(),
  trendData: z.array(z.any()),
  dataPoints: z.array(z.any()),
  hover: z.record(z.string(), z.any()),
});

export const LeaderboardRowSchema = z.object({
  profileId: z.string(),
  firstName: z.string(),
  lastName: z.string(),
  metrics: z.object({
    totalAttempts: LeaderboardMetricSchema,
    highestScoreAvg: LeaderboardMetricSchema,
    messagesPerSession: LeaderboardMetricSchema,
    personaResponseSeconds: LeaderboardMetricSchema,
    timeSpentMinutes: LeaderboardMetricSchema,
    improvementRatePerDay: LeaderboardMetricSchema,
    perfectScoreCount: LeaderboardMetricSchema,
    quickestPassMinutes: LeaderboardMetricSchema,
  }),
});

export const LeaderboardBundleResponseSchema = z.object({
  data: z.array(LeaderboardRowSchema),
});

export type LeaderboardMetric = z.infer<typeof LeaderboardMetricSchema>;
export type LeaderboardRow = z.infer<typeof LeaderboardRowSchema>;
export type LeaderboardBundleResponse = z.infer<
  typeof LeaderboardBundleResponseSchema
>;
